include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,266 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_AUTO_TUNE_POLICY_H
+#define _ONEDPL_AUTO_TUNE_POLICY_H
+
+#include <stdexcept>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <chrono>
+#include <limits>
+#include <vector>
+#include <type_traits>
+#include <tuple>
+#include "oneapi/dpl/internal/dynamic_selection_traits.h"
+#if _DS_BACKEND_SYCL != 0
+    #include "oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h"
+#endif
+namespace oneapi {
+namespace dpl {
+namespace experimental {
+#if _DS_BACKEND_SYCL != 0
+  template <typename Backend=sycl_backend, typename... KeyArgs>
+#else
+  template <typename Backend, typename... KeyArgs>
+#endif
+  class auto_tune_policy {
+
+    static constexpr double never_resample = 0.0;
+    static constexpr int use_best_resource = -1;
+
+    using wrapped_resource_t = typename std::decay_t<Backend>::execution_resource_t;
+    using size_type = typename std::vector<typename Backend::resource_type>::size_type;
+
+    using timing_t = uint64_t;
+
+    struct time_data_t {
+      uint64_t num_timings_ = 0;
+      timing_t value_ = 0;
+    };
+
+    struct tuner_t {
+      std::mutex m_;
+
+      std::chrono::steady_clock::time_point t0_;
+
+      timing_t best_timing_ = std::numeric_limits<timing_t>::max();
+      wrapped_resource_t best_resource_;
+
+      const size_type max_resource_to_profile_;
+      uint64_t next_resource_to_profile_ = 0;
+
+      using time_t = std::map<size_type, time_data_t>;
+      time_t time_;
+
+      double resample_time_ = 0.0;
+
+      tuner_t(wrapped_resource_t br, size_type resources_size, double rt)
+        : t0_(std::chrono::steady_clock::now()),
+          best_resource_(br),
+          max_resource_to_profile_(resources_size),
+          resample_time_(rt) {}
+
+      size_type get_resource_to_profile() {
+        std::lock_guard<std::mutex> l(m_);
+        if (next_resource_to_profile_ < 2*max_resource_to_profile_) {
+          // do everything twice
+          return next_resource_to_profile_++ % max_resource_to_profile_;
+        } else if (resample_time_ == never_resample) {
+          return use_best_resource;
+        } else {
+          auto now = std::chrono::steady_clock::now();
+          auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now-t0_).count();
+          if (ms < resample_time_) {
+            return use_best_resource;
+          } else {
+            t0_ = now;
+            return next_resource_to_profile_ = 0;
+          }
+        }
+      }
+
+      // called to add new profile info
+      void add_new_timing(wrapped_resource_t r, timing_t t) {
+        std::unique_lock<std::mutex> l(m_);
+        timing_t new_value = t;
+        if (time_.count(0) == 0) {
+          // ignore the 1st timing to cover for JIT compilation
+          time_[0] = time_data_t{0, std::numeric_limits<timing_t>::max()};
```

Comment:
Do I get it right that of the whole the `std::map` created at lines 61-61 you only ever use a single element `time_[0]`?

Reply:
This was a bug introduced yesterday by removal of offsets in policies. auto_tune did not support offsets, but had a similarly named variable that held the index into the set of resources. This index was stored in the selection (since the selection would be for a particular resource) but was removed in error. Each element in time_ stores the timing for the resource at that index in the set of resources.  We are going to revert these changes to auto_tuned.